Sprint 6 - final task.

imposmas opened the pull request.

imposmas committed:
Sprint 6 - final task.

imposmas committed:
Sprint 6 - final task.
Correct variables based on checkstyle

imposmas committed:
Sprint 6 - final task.
Correct variables based on checkstyle subtasks objects in Main

imposmas committed:
Sprint 6 - final task.
checkstyle correction in Epic

imposmas committed:
Sprint 6 - final task.
checkstyle correction in all code

imposmas committed:
Sprint 6 - final task.
InMemoryHistoryManagerTest tasks correction based on changes in task storage

Fleaman777 commented on src/ru/yandex/practicum/kanban/managers/InMemoryHistoryManager.java:
```diff
@@ -3,24 +3,80 @@
-public class InMemoryHistoryManager implements HistoryManager{
+public class InMemoryHistoryManager implements HistoryManager {
+
+    HashMap<Integer, Node> tasksHistory = new HashMap<>();
```
Лучше сделать поля private, чтобы обеспечить инкапсуляцию.

Fleaman777 commented on src/ru/yandex/practicum/kanban/managers/InMemoryHistoryManager.java:
```diff
@@ -3,24 +3,80 @@
+            after.prev = before;
+        }
+        tasksHistory.remove(node.item.getId());
+    }
```
Пропущена обработка связей в случае если after == null. Добавь after.prev = before в первый блок и before.next = after во второй.

Fleaman777 commented on src/ru/yandex/practicum/kanban/managers/InMemoryHistoryManager.java:
```diff
@@ -3,24 +3,80 @@
+        tasksHistory.remove(node.item.getId());
+    }
+
+    private static class Node {
```
отличное решение, сделать этот класс внутренним

Fleaman777 commented on test/ru/yandex/practicum/kanban/managers/InMemoryHistoryManagerTest.java:
```diff
@@ -35,32 +41,32 @@ void addAndGetHistoryTest() {
-        historyManager.addToHistory(task);
-        historyManager.addToHistory(task);
-        historyManager.addToHistory(task);
+    void HistoryDoesNotStoreDuplicatesTest() {
```
по джава код стайлу: методы должны начинаться с маленькой буквы

Fleaman777 commented on test/ru/yandex/practicum/kanban/managers/InMemoryTaskManagerTest.java:
```diff
@@ -97,28 +95,39 @@ void taskIsEqualsAfterAddToManager(){
+    @Test
+    void EpicDoesNotStoreDeletedSubtasks() {
+        taskManager.deleteSubTask(SubTaskId2);
+        assertTrue(!taskManager.getEpicSubTasks(epicId1).contains(SubTaskId2));
```
вижу. что новые тесты добавлены, это хорошо. 
но посмотри еще раз задание, какие кейсы необходимо покрыть

imposmas committed:
Sprint 6 - final task.
- make setters 'private'
- java style correction
- correct logic of Nodes calculation
- Tests correction based on private setters (objects update by object re-creation via appropriate constructor)
- change updateSuTask to avoid cases with non-existing epicIds (Удаляемые подзадачи не должны хранить внутри себя старые id)
- EpicDoesNotStoreDeletedSubtasks should be covered (Внутри эпиков не должно оставаться неактуальных id подзадач)
- private setters are covered by new constructors

imposmas committed:
Sprint 6 - final task.
- updateTaskStatusTest() change

imposmas commented:
если с тестами все еще в пролете, то подскажи пожалуйста как будет правильно реализовать их 
дохожу до того, что не могу создать ломаные объекты чтоб потом на них какое либо действие проверить, потому что везде куча проверок на соотвесвие....

VadimZharkov commented on src/ru/yandex/practicum/kanban/managers/InMemoryTaskManager.java:
```diff
@@ -21,94 +21,93 @@ public class InMemoryTaskManager<T extends Task> implements TasksManager {
 
```
HistoryManager historyManager - модификатор доступа?

VadimZharkov commented on src/ru/yandex/practicum/kanban/managers/InMemoryHistoryManager.java:
```diff
@@ -3,24 +3,97 @@
-        if(tasksHistory.size() == HISTORY_SIZE){
-            tasksHistory.remove(0);
+    public void remove(int id) {
+        if (tasksHistory.get(id) != null) {
```
Лучше за одно обращение к мапе
```
Node node = tasksHistory.get(task.getId());
if (node != null) {
      removeNode(node);
}
```

VadimZharkov reviewed (changes requested):
Можно улучшить

imposmas committed:
Sprint 6 - final task.
- make history manager private in TaskManager
- correct node removal logic
